src/stream/src/executor/backfill/cdc/upstream_table/snapshot.rs

```diff
@@ -100,36 +101,63 @@ impl UpstreamTableRead for UpstreamTableReader<ExternalStorageTable> {
             self.inner.schema_table_name(),
             args.current_pos,
             primary_keys,
-            limit,
+            read_limit,
         );
 
         pin_mut!(row_stream);
 
-        let mut builder = DataChunkBuilder::new(self.inner.schema().data_types(), args.chunk_size);
+        let mut builder = DataChunkBuilder::new(
+            self.inner.schema().data_types(),
+            limited_chunk_size(args.rate_limit_rps),
+        );
         let chunk_stream = iter_chunks(row_stream, &mut builder);
 
-        if args.chunk_size == 0 {
+        if args.rate_limit_rps == Some(0) {
             // If limit is 0, we should not read any data from the upstream table.
             // Keep waiting util the stream is rebuilt.
             let future = futures::future::pending::<()>();
             future.await;
+            unreachable!();
         }
-        let limiter = {
-            let quota = Quota::per_second(NonZeroU32::new(args.chunk_size as u32).unwrap());
-            let clock = MonotonicClock;
-            RateLimiter::direct_with_clock(quota, &clock)
-        };
+
+        let limiter = args.rate_limit_rps.map(|limit| {
+            tracing::info!(rate_limit = limit, "rate limit applied");
+            RateLimiter::direct_with_clock(
+                Quota::per_second(NonZeroU32::new(limit).unwrap()),
+                &MonotonicClock,
+            )
+        });
+
         #[for_await]
         for chunk in chunk_stream {
             let chunk = chunk?;
-            if chunk.cardinality() != 0 {
-                limiter
-                    .until_n_ready(NonZeroU32::new(chunk.cardinality() as u32).unwrap())
-                    .await
-                    .unwrap();
+            let cardinality = chunk.cardinality();
+
+            if args.rate_limit_rps.is_none() || cardinality == 0 {
+                // no limit, or empty chunk
+                yield Some(chunk);
+                continue;
+            }
+
+            // Apply rate limit, see `risingwave_stream::executor::source::apply_rate_limit` for more.
+            // May be should be refactored to a common function later.
+            let limiter = limiter.as_ref().unwrap();
+            let limit = args.rate_limit_rps.unwrap();
+            if cardinality <= limit as usize {
+                let n = NonZeroU32::new(cardinality as u32).unwrap();
+                // `InsufficientCapacity` should never happen because we have check the cardinality
+                limiter.until_n_ready(n).await.unwrap();
+                yield Some(chunk);
+            } else {
+                // Cut the chunk into smaller chunks
+                for chunk in chunk.split(limit as usize) {
```

Comment:
Do we need to split chunk here? In the code above, the `DataChunkBuilder` is already initialized with rate_limit as the chunk size.

Reply:
You're right. Fixed.